### PR TITLE
[flink] [hive] Introduce procedure to migrate table from hive to paimon

### PR DESCRIPTION
**Pull request by leaves12138**
### Purpose

Introduce procedure to migrate table from hive to paimon 
Or
Add file from hive to paimon.

We should use `FlinkGenericCatalog` to migrate table.

***CALL <font color="green">sys.migrate_table</font>(&#39;hive&#39;, &#39;&lt;hive_database&gt;.&lt;hive_tablename&gt;&#39;, &#39;&lt;paimon_tableconf&gt;&#39;);***

***CALL <font color="green">sys.migrate_file</font>(&#39;hive&#39;, &#39;&lt;hive_database&gt;.&lt;hive_table_name&gt;&#39;, &#39;&lt;paimon_database&gt;.&lt;paimon_tablename&gt;&#39;);***

Example:
```sql
CREATE CATALOG PAIMON WITH ('type'='paimon-generic', 'hive-conf-dir' = '/path/to/hive-site.xml');

USE CATALOG PAIMON;

CALL sys.migrate_table('default.hivetable', 'file.format=orc');
```

```sql
CREATE CATALOG PAIMON WITH ('type'='paimon-generic', 'hive-conf-dir' = '/path/to/hive-site.xml');

USE CATALOG PAIMON;

CALL sys.migrate_file('default.hivetable', 'default.paimontable');
```
### Tests

IT case added.

### Documentation

Document added

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/migrate/DataTypeWriter.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import org.apache.paimon.types.TimeType;
+import org.apache.paimon.types.TimestampType;
+import org.apache.paimon.types.TinyIntType;
+import org.apache.paimon.types.VarBinaryType;
+import org.apache.paimon.types.VarCharType;
+
+import java.math.BigDecimal;
+
+/** Generate different converter to write data. */
+public class DataTypeWriter implements DataTypeVisitor<DataConverter> {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/migrate/DataTypeWriter.java:49]: Can you use `TypeUtils.castFromString` and `BinaryWriter.createValueSetter`?

File: paimon-core/src/main/java/org/apache/paimon/operation/AppendOnlyFileStoreRead.java [hunk 2]
```diff
@@ -104,7 +104,7 @@ public RecordReader<InternalRow> createReader(DataSplit split) throws IOExceptio
         DataFilePathFactory dataFilePathFactory =
                 pathFactory.createDataFilePathFactory(split.partition(), split.bucket());
         List<ConcatRecordReader.ReaderSupplier<InternalRow>> suppliers = new ArrayList<>();
-        if (split.beforeFiles().size() > 0) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AppendOnlyFileStoreRead.java:107]: Here don't need to use isEmpty, size > 0 is also a good way.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/MigrateTableProcedure.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    public String[] call(
+            ProcedureContext procedureContext, String sourceTablePath, String properties)
+            throws Exception {
+        TableEnvironmentImpl tableEnvironment =
+                TableEnvironmentImpl.create(EnvironmentSettings.inBatchMode());
+        Identifier sourceTableId = Identifier.fromString(sourceTablePath);
+
+        CatalogTable sourceFlinkTable =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/MigrateTableProcedure.java:62]: move these logical into paimon-hive, we don't need to rely on Flink Catalog APIs.

File: docs/content/migration/migration-from-hive.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+under the License.
+-->
+
+# Hive Table Migration
+
+Apache Hive supports ORC, Parquet file formats that could be migrated to Paimon. 
+When migrating data to a paimon table, the origin table will be permanently disappeared. So please back up your data if you
+still need the original table. The migrated table will be [unware-bucket append-only table]({{< ref "concepts/append-only-table#append-for-scalable-table" >}}).
+
+Now, we can use flink generic catalog with Migrate Table Procedure and Migrate File Procedure to totally migrate a table from hive to paimon.
```
Comment (JingsongLi) [inline docs/content/migration/migration-from-hive.md:33]: We should try to not limit by flink generic catalog, this feature can be engine unrelated.

**Review by JingsongLi — CHANGES_REQUESTED** [review]

File: docs/content/migration/migration-from-hive.md (hunk 4 above)
Comment (leaves12138) [reply docs/content/migration/migration-from-hive.md:33]: done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/MigrateTableProcedure.java (hunk 3 above)
Comment (leaves12138) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/MigrateTableProcedure.java]: done

File: paimon-core/src/main/java/org/apache/paimon/operation/AppendOnlyFileStoreRead.java (hunk 2 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/operation/AppendOnlyFileStoreRead.java]: reverted

File: paimon-core/src/main/java/org/apache/paimon/migrate/DataTypeWriter.java (hunk 1 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/migrate/DataTypeWriter.java]: done

File: paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/MigrationTableProcedure.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+package org.apache.paimon.spark.procedure;
+
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.types.StructType;
+
+/** Test. */
+public class MigrationTableProcedure extends BaseProcedure {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/MigrationTableProcedure.java:26]: Remove this

File: docs/content/migration/migration-from-hive.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+Apache Hive supports ORC, Parquet file formats that could be migrated to Paimon. 
+When migrating data to a paimon table, the origin table will be permanently disappeared. So please back up your data if you
+still need the original table. The migrated table will be [unaware-bucket append-only table]({{< ref "concepts/append-only-table#append-for-scalable-table" >}}).
+
+Now, we can use paimon hive catalog with Migrate Table Procedure and Migrate File Procedure to totally migrate a table from hive to paimon.
+
+* Migrate Table Procedure: Paimon table does not exist, use the procedure upgrade hive table to paimon table. Hive table will disappear after action done.
+* Migrate File Procedure:  Paimon table already exists, use the procedure to migrate files from hive table to paimon table. **Notice that, Hive table will also disappear after action done.**
+
+<span style="color: red; "> **We highly recommend to back up hive table data before migrating, because migrating action is not atomic. If been interrupted while migrating, you may lose your data.** </span>
```
Comment (JingsongLi) [inline docs/content/migration/migration-from-hive.md:38]: Document file format.

File: paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/MigrationTableProcedure.java (hunk 5 above)
Comment (leaves12138) [reply paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/MigrationTableProcedure.java:26]: done

File: paimon-flink/paimon-flink-common/pom.xml [hunk 7]
```diff
@@ -120,6 +121,20 @@ under the License.
             </exclusions>
         </dependency>
 
+
+        <dependency>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-metastore</artifactId>
```
[the remark this answers is not in the document]
Comment (leaves12138) [reply paimon-flink/paimon-flink-common/pom.xml:127]: done

File: docs/content/migration/migration-from-hive.md (hunk 6 above)
Comment (leaves12138) [reply docs/content/migration/migration-from-hive.md:41]: done

File: paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/MigrationTableProcedure.java (hunk 5 above)
Comment (leaves12138) [reply paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/MigrationTableProcedure.java]: done

File: paimon-flink/paimon-flink-common/pom.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (leaves12138) [reply paimon-flink/paimon-flink-common/pom.xml]: done

**Comment by leaves12138** [comment]
Added Migrate Table Action and its test to project

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/migrate/HiveMigrator.java [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+
+        if (sync) {
+            List<CommitMessage> commitMessages = new ArrayList<>();
+            tasks.forEach(task -> commitMessages.add(task.get()));
+            paimonTable.newBatchWriteBuilder().newCommit().commit(commitMessages);
+        } else {
+            Queue<CommitMessage> commitMessages = new LinkedBlockingQueue<>();
+            List<Future<?>> futures = new ArrayList<>();
+            ExecutorService executors =
+                    Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
```
Comment (JingsongLi) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/migrate/HiveMigrator.java:133]: Use `COMMON_IO_FORK_JOIN_POOL`?

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/migrate/HiveMigrator.java [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+        List<MigrateTask> tasks = new ArrayList<>();
+        if (partitionsNames.isEmpty()) {
+            tasks.add(importUnPartitionedTableTask(fileIO, sourceHiveTable, paimonTable));
+        } else {
+            tasks.addAll(
+                    importPartitionedTableTask(
+                            client, fileIO, partitionsNames, sourceHiveTable, paimonTable));
+        }
+
+        if (sync) {
```
Comment (JingsongLi) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/migrate/HiveMigrator.java:125]: Why we need a sync mode? Actually, single thread mode.

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/migrate/HiveMigrator.java (hunk 8 above)
Comment (leaves12138) [reply paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/migrate/HiveMigrator.java:133]: done

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/migrate/HiveMigrator.java (hunk 9 above)
Comment (leaves12138) [reply paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/migrate/HiveMigrator.java:125]: Reasonable, we don't need single thread mode. At first, I want to return a Future<MigrateResult>, but seems it is not necessary.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**